### PR TITLE
Replace Mono.Addins with our own addin implementation

### PR DESCRIPTION
**Pull request by CharliePoole**
** Now Ready to Review and Merge**

I'm creating this PR as I _begin_ work on issue #749, rather than at the end, so that we can use it as a place to discuss how we want our addins to work as well as for normal code review.

The initial commit simply hides all interaction with Mono.Addins behind a new service, which I called ExtensionService - not very original I suppose.

I have also created a rough spec for this implementation, which you can find here:
    https://github.com/nunit/dev/wiki/Engine-Addins

Let's have any discussion of either the spec or the code right on this PR.

### CODE REVIEW

**Comment by CharliePoole** [comment]
Rebased after Cecil use was merged into engine.

**Comment by rprouse** [comment]
Not much to comment on in the code since it is just laying the groundwork to continue. I've read through the spec and it is pretty much in line with what I was thinking. The only thing that I might do differently is to have an assembly level attribute to determine what interfaces the plugin provides. You say that we won't be scanning very many assemblies, but we do need to scan every type in every assembly don't we? An assembly level attribute with two parameters, the interface and the implementation could really speed this up.

Also, will you be using Cecil for this? If it is as quick as they say it is, we may not need assembly attributes :smile: 

Where will assemblies that are referenced by the extensions go? I assume alongside the extension in the addins directory?

**Comment by CharliePoole** [comment]
I started experimenting and have changed a few things - I'll bring the spec
up to date so it can serve as documentation later.

I'm now using an assembly-level attribute to identify extension points. It
would be possible to do the same for addins, although I am intrigued by the
idea of simply putting an ExtensionAttribute on a class and letting it be
found that way. I'll experiment with it both ways.

I think I should use Cecil for everything except reading the engine's own
assembly-level attributes for extension points. Aside from efficiency
considerations, it removes the need to unload assemblies that don't have
any addins. We could just build a list of assemblies to be loaded for each
extension type and only really load it when needed. I'll try this out.

One compromise I have thought of: for known addins, specifically called out
in the addins.xml file, we might scan the entire assembly. Assemblies
simply found in a directory would not be scanned without an assembly-level
attribute. Again, I'll experiment.

I should have a new commit and a spec update in a few days.

**Comment by CharliePoole** [comment]
I updated the spec at https://github.com/nunit/dev/wiki/Engine-Addins to reflect my current thinking based on some spikes I have run.

**Comment by CharliePoole** [comment]
The second commit adds attributes used to identify extension points and extensions and tests that they are found. This is about as far as I can easily go without removing Mono.Addins because of naming conflicts. Next commit will do that.

**Comment by CharliePoole** [comment]
Commit 311f8c0 removes Mono.Addins entirely from the build and replaces it with a rudimentary addin implementation as described in the spec.

What's implemented:
* Extension Points and Extensions
* Addin assemblies located in the addins directory
* All our existing extensions, previously implemented with Mono.Addins.
* Existing tests all pass.

Not implemented:
* Loading addins from arbitrary directories outside of the ApplicationBase.
* Addins (as defined in the spec)
* Extensions to extensions
* Non-happy-path tests including bad extensions, invalid paths, etc.

Note that the core engine still does not support extensions. We may decide to do this as a next step. OTOH, having a low-overhead extensibility implementation could actually remove the need for a separate core engine if we decided to put all the non-core functionality into extensions.

At this point, I think merging what we have and creating separate issues for any further enhancements is probably the best approach.

**Comment by CharliePoole** [comment]
@nunit/core-team This is now ready for review and possible merging.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+        public IEnumerable<ExtensionNode> Extensions
+        {
+            get { return _extensions;  }
+        }
+
+        #endregion
+
+        public ExtensionPoint GetExtensionPoint(string path)
```
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: Everywhere we get extensions, we do it based on the interface, not the path, so this seems unused and unneeded. By extension, I don't think the path is needed in the ExtensionAttribute. If anything, the attribute should specify the interface that an extension is providing an extension for. Hard coded string paths seem cryptic and error prone to me.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                foreach (var file in addinsFiles)
+                    ProcessAddinsFile(startDir.FullName, file.FullName);
+            else
+                foreach (var file in startDir.GetFiles("*.dll"))
+                    FindExtensionsInAssembly(file.FullName);
+        }
+
+        private void ProcessAddinsFile(string baseDir, string fileName)
+        {
+            var doc = new XmlDocument();
```
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: Using XML for this file seems like overkill for a list of addin directories. Wouldn't it be easier to just have a flat file, one addin directory per line? We could ignore blank lines and lines starting with #. Were you planning on including other information in the addins files?

Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: I planned on a flat file (see spec) but then adjusted so I could use the
pre-existing file from mono.addins. Their format has directory entries and
patterns/names of assemblies to ignore when scanning. Up to now, I have not
felt the need for the ignored files.

My own plan was to have lines for directories and lines for files, which
might include wild-cards. It seems like it may be easier to just indicate
which lines are directories and which are files rather than try to figure
it out. We could use a prefix like D: or A: for directories and assemblies,
but XML seemed pretty easy for this.

Bottom line: I think we can go either way. What do you think, considering
the possible addition of individual assemblies and bearing in mind that
users may edit this file.

Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: Personally, I think that if users are editing the file, all the more reason to keep it to a minimum and stay away from XML.

To distinguish between files and directories, we could just stipulate that directories must end in a slash? If we want to ignore files, those lines could just be preceeded by a minus sign.

We could also use a YAML format, something like;

```
include:
    directory1\
    directory2\
exclude:
    directory1\file1.dll
```

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: Currently, we use the Path on the extension to find the extension point with which it is associated. My original idea was to leave path off extensions where possible, but I ran into some problems dealing with edge cases. I want to go back to this. I think you should be able to just use [Extension] in all the cases we currently support.

You can get extensions by interface (as we do now) provided there is only one extension point that would satisfy that interface. The current implementation will break if two extension points take the same interface, but in principle, it should be possible to do this. For example, I can imagine two extension points that returned IEnumerable<string> with the strings used for different purposes. So, at least for some edge cases, we need the path.

Strings can be bad but they are our general way of communicating among components that are not built together.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            var module = AssemblyDefinition.ReadAssembly(assemblyName).MainModule;
+            foreach (var type in module.GetTypes())
+                foreach (var attr in type.CustomAttributes)
+                    if (attr.AttributeType.FullName == "NUnit.Engine.Extensibility.ExtensionAttribute")
+                    {
+                        var node = new ExtensionNode(assemblyName, type.FullName);
+                        node.Path = (string)attr.ConstructorArguments[0].Value;
+                        _extensions.Add(node);
+
+                        var ep = GetExtensionPoint(node.Path);
```
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: As I stated before, I would drop the path and just use the interfaces that they are extending.

Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: How would you deal with duplicate interfaces? I think we can minimize the use of the path, but not eliminate it unless we say that no two extension points can accept the same interface. That seems arbitrary and could cause trouble down the line.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 1 above)
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: I'll give you the benefit of the doubt here, but it feels over-engineered :smile:

We could get around ever needing a path if we add the constraint that every extension point needs to expect a unique interface. Your two IEnumerable<string> extension points would need to create a unique interface like;

```C#
public interface IExtensionOne : IEnumerable<string> {}
public interface IExtensionTwo : IEnumerable<string> {}
```

Personally, I doubt we will do this sort of thing very often, so the constraint would simplify development in the majority of cases and nobody would ever have to know magic strings, just the interface they are implementing.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: Do you know of a YAML parsing class I can borrow?

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: It's more advance engineering than I usually do, for sure. But that comes from working on something that shouldn't be changed once implemented. Or at least not changed in a breaking way.

Note that the rule would not be simply "Every extension point needs to expect a unique interface." To that, you have to add the rule "An extension must implement the interface required by it's extension point an may not implement any interface required by any **other** extension point." It seems to me that this puts an extreme burden on the authors of extensions, who can't really know what other extension points may be available in a given system. 

That said, I'll do some more work on this to see what I can come up with.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 2 above)
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: There is YamlDotNet, http://aaubry.net/pages/yamldotnet.html, but it is fairly heavy for what we need. We could probably write a simple one that satisfies the subset of YAML we need;

1. Split lines at #, ignore second part (comments)
2. Ignore blank or empty lines
3. Lines starting with no whitespace and ending in :, switch state to adding to include, exclude, etc.
4. Lines starting with whitespace, trim whitespace and add to include, exclude, etc depending on state.
5. Anything else is an error

This will only allow for labels and items, but will probably do what we need?

Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: Simple enough. I'll give it a shot and see about replacing the mono.addins
file with our own. Should we call it .yml, even though it's a subset?

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 1 above)
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: > An extension must implement the interface required by it's extension point an may not implement any interface required by any other extension point.

Not if the `ExtensionAttribute` took a `Type` parameter for the interface they are implementing.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 2 above)
Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: I think we just call it a .addin since it is a subset and we made up the format 'based' on YAML.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: True. But I really like the fact that you can just use [Extension] in Mono.Addins. Still, Type could be optional.

Comment (rprouse) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: OK, then Type is only required if you implement multiple interfaces and one might conflict. Personally, I think that would be extremely rare. :smile:

**Comment by CharliePoole** [comment]
The latest commit keeps Path but makes it optional. We don't actually use it in any of our extensions.

Just in case it's not clear, the Path is simply an identifier, not a file path. It should never need to be updated unless we arbitrarily change the identifier. The name Path is a legacy of Mono.Addins and reflects that this is a path in the tree of ExtensionPoints.

This has some error checking now, but the tests are still all happy paths. I think that we need more complete testing of the extension mechanism, but that's an integration test. Once we decide what to do about integration testing, we can create some tests using fake addins.

Still need to do some work on this PR, in particular as regards to the .addins file.

**Comment by CharliePoole** [comment]
I should add that the code doesn't yet handle extensions with duplicate interfaces yet. The first extension point found will be used. It's not needed now and probably won't be until users start writing extensions. Once we have a better test harness, I'll start throwing when ambiguous entries are detected.

Note too that I do throw on errors. This seems to make sense, since the application detects extension points and extensions on startup. If an extension causes a crash, then it simply has to be removed. An alternative would be to proceed without it, but it's not clear how the user should be notified in that case.

File: src/NUnitEngine/nunit.engine/Services/ExtensionService.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitEngine/nunit.engine/Services/ExtensionService.cs]: I've experimented with this, and I think I prefer to stick with XML, because

1. It takes 2 to 3 times the code to parse even the simplified flat file if you check for errors.

2. It creates another format to deal with, albeit simple.

3. We already use XML for just about everything else: representing tests and results, filters, progress reports, NUnit settings, format of nunit and VS projects.

4. Anyone familiar with Mono.Addins already creates files like this in XML.

**Comment by CharliePoole** [comment]
For the new commit, I have gone with a purely flat file. The only structure is that comments using '#' are permitted. Each line contains a directory or a file pattern. File patterns may use '*' but directories may not. We identify a directory based on the fact that it exists.

This seems to work well. I have tested with a directory name (addins), a pattern (addins/*.dll) and a list of actual extension assemblies. The last is what's currently in the .addins file.

For now, I left the code for the XML file in the code using conditional compliation so we can easily compare the two implementations.

**Comment by rprouse** [comment]
I prefer the flat format. Wherever possible, I think we should keep these configuration files as simple as possible. If in the future we want to add exclusions, we could start the line with a minus sign.

**Comment by rprouse** [comment]
@CharliePoole, are you still okay to merge this? I am good.

**Comment by CharliePoole** [comment]
I'll remove the unused code, check over everything a last time and merge it.

**Comment by CharliePoole** [comment]
I made a few changes in the last commit:

1. Comments on some methods
2. Remove unused code
3. Uncommented and modified ExtensionPoint.Install and Remove programs, adding a bit of error checking to install. Initialization now uses Install rather than directly adding to the extension point.
4. Modified several methods to return an IEnumerable rather than an array, allowing use of yield.

The changes are basic and the tests all continue to pass, so I'm merging.
